nselib/punycode.lua

```diff
@@ -0,0 +1,664 @@
+          local ans = digitToBasic(t + qMinusT % baseMinusT, 0)
+
+          table.insert(output, char(ans))
+
+          q = floor(qMinusT / baseMinusT)
+
+          k = k + base
+        until false
+
+        local ans = digitToBasic(q, 0)
+        table.insert(output, char(ans))
+        bias = adapt(delta, handledCPCountPlusOne, handledCPCount == basicLength)
+
+        delta = 0
+        handledCPCount = handledCPCount + 1
+      end
+    end
+
+    delta = delta + 1
+    n = n + 1
+
+  end
+
+  return table.concat(output, '')
+
+end
+
+-- This function converts a Punycode string of ASCII-only symbols to a
+-- string of Unicode symbols.
+--
+-- @param input The Punycode string of ASCII-only symbols.
+-- @param encoder Defines the type of encoding format to be used.
+-- @return The resulting string of Unicode symbols.
+-- The following function is adapted from punycode.js by Mathias Bynens
+-- under the MIT License.
+function decode_input(input, encoder)
+
+  local output = {}
+  local inputLength = #input
+  local i = 0
+  local n = initialN
+  local bias = initialBias
+
+  local basic
+  if find(reverse(input), delimiter) then
+    basic = #input - find(reverse(input), delimiter)
+  else
+    basic = -1
+  end
+
+  if basic < 0 then
+    basic = 0
+  end
+
+  for j = 1, basic do
+    local c = sub(input, j, j)
+    local value = byte(c)
+
+    if value >= 0x80 then
+      --error("Not basic")
+      return nil, "Not basic exception occurred."
+    end
+    table.insert(output, value)
+  end
+
+  local index
+  if basic > 0 then
+    index = basic + 1
+  else
+    index = 0
+  end
+
+  while index < inputLength do
+    local oldi = i
+    local w = 1
+    local k = base
+
+    repeat
+
+      if index >= inputLength then
+        --error("Invalid input")
+        return nil, "Invalid input exception occurred."
+      end
+
+      local c = sub(input, index+1, index+1)
+      local value = byte(c)
+      local digit = basicToDigit(value)
+
+      index = index + 1
+
+      if (digit >= base or digit > floor((maxInt - i) / w)) then
+        --error('overflow');
+        return nil, "Overflow exception occurred."
+      end
+      i = i + digit * w;
+
+      local t
+      if k <= bias then
+        t = tMin
+      else
+        if k >= bias + tMax then
+          t = tMax
+        else
+          t = k - bias
+        end
+      end
+
+      if digit < t then
+        break
+      end
+
+      local baseMinusT = base - t;
+      if (w > floor(maxInt / baseMinusT)) then
+        --error('overflow');
+        return nil, "Overflow exception occurred."
+      end
+
+      w = w * baseMinusT;
+      k = k + base
+
+    until false
+
+    local out = #output + 1;
+
+    bias = adapt(i - oldi, out, oldi == 0)
+
+    -- `i` was supposed to wrap around from `out` to `0`,
+    -- incrementing `n` each time, so we'll fix that now:
+    if (floor(i / out) > maxInt - n) then
+      --error('overflow');
+      return nil, "Overflow exception occurred."
```

Comment:
Are we checking for this and other punycode errors when we call these functions from `idna.lua`? We should be doing so and passing the error up to the caller.

Reply:
`punycode.mapLabels` is the only function being used in `idna.lua` and this function is returned to the caller, so if there is an error the error will be returned to the caller along with `nil` as response value.

Reply:
So, effectively it becomes the job of the caller to handle the errors being thrown.